rustworkx-core/src/coloring.rs

```diff
@@ -220,8 +221,218 @@ where
     edge_colors
 }
 
-#[cfg(test)]
+struct MisraGries<G: GraphBase> {
+    // The input graph
+    graph: G,
+    // Partially assigned colors (indexed by internal edge index)
+    colors: Vec<Option<usize>>,
+}
+
+impl<G> MisraGries<G>
+where
+    G: EdgeIndexable + IntoEdges,
+{
+    pub fn new(graph: G) -> Self {
+        let colors = vec![None; graph.edge_bound()];
+        MisraGries { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: G::NodeId) -> HashSet<usize> {
+        let used_colors: HashSet<usize> = self
+            .graph
+            .edges(u)
+            .filter_map(|edge| self.colors[EdgeIndexable::to_index(&self.graph, edge.id())])
+            .collect();
+        used_colors
+    }
+
+    // Returns the smallest free (aka unused) color at node u
+    fn get_free_color(&self, u: G::NodeId) -> usize {
+        let used_colors = self.get_used_colors(u);
+        let free_color: usize = (0..)
+            .position(|color| !used_colors.contains(&color))
+            .unwrap();
+        free_color
+    }
+
+    // Returns true iff color c is free at node u
+    fn is_free_color(&self, u: G::NodeId, c: usize) -> bool {
+        let used_colors = self.get_used_colors(u);
+        !used_colors.contains(&c)
```

Comment:
Could you just do `self.colors.contains(&Some(c))` (or something like this, I might have got the position of `&` incorrect)? You'd have the same linear overhead. 

More generally though I'm wondering if it makes more sense to store a `HashSet` of used colors in the `MisraGries` struct so you don't have to generate it on the fly every time. Then you could make this a O(1) lookup every time. It would require doing book keeping on adding or changing colors but it should be manageable.

Reply:
I have followed the above suggestion in [4e22ad7](https://github.com/Qiskit/rustworkx/pull/902/commits/4e22ad7abdb6c27eac28c0ea88927d94ea50eb93), and indeed it gives a significant speedup on medium/large-sized complete graphs. E.g., on my laptop previously it took `16.95` seconds to color a complete graph over 90 vertices, and now it takes `0.58` seconds. For heavy-hex graphs the difference is less pronounced, but the new implementation is still a bit faster.